link_shortener/form_routes.py

```diff
@@ -209,14 +219,72 @@ async def update_link_save(request, user, status, link_id):
     try:
         async with request.app.engine.acquire() as conn:
             trans = await conn.begin()
-            await conn.execute(
-                table.update().where(
-                    table.columns['id'] == link_id
-                ).values(
-                    password=form.password.data,
-                    url=form.url.data
+            if form.password.data:
+                link_query = await conn.execute(
+                    table.select().where(
+                        table.columns['id'] == link_id
+                    )
                 )
-            )
+                link_data = await link_query.fetchone()
+                if link_data.password:
+                    fresh_salt = os.urandom(32)
+                    password = hashlib.pbkdf2_hmac(
+                        'sha256',
+                        form.password.data.encode('utf-8'),
+                        fresh_salt,
+                        100000
+                    )
+                    await conn.execute(
+                        table.update().where(
+                            table.columns['id'] == link_id
+                        ).values(
+                            url=form.url.data,
+                            password=password,
+                            switch_date=form.switch_date.data
+                        )
+                    )
+                    await conn.execute(
+                        salts.update().where(
+                            salts.columns['identifier'] == link_data.identifier
+                        ).values(
+                            salt=fresh_salt
+                        )
+                    )
+
+                else:
+                    salt = os.urandom(32)
+                    password = hashlib.pbkdf2_hmac(
+                        'sha256',
+                        form.password.data.encode('utf-8'),
+                        salt,
+                        100000
+                    )
+                    await conn.execute(
+                        table.update().where(
+                            table.columns['id'] == link_id
+                        ).values(
+                            url=form.url.data,
+                            password=password,
+                            switch_date=form.switch_date.data
+                        )
+                    )
+                    await conn.execute(
+                        salts.insert().values(
+                            identifier=link_data.identifier,
+                            salt=salt
+                        )
+                    )
+
+            else:
+                await conn.execute(
+                    table.update().where(
+                        table.columns['id'] == link_id
+                    ).values(
+                        url=form.url.data,
+                        switch_date=form.switch_date.data
+                    )
+                )
+
```

Comment:
I was hoping it'd get better once I've written methods for the individual blocks/executions. As for the structure, I don't know how to make it nicer.

Reply:
I think something like this might be more readable:
```python
link_update = table.update()
    .where(table.columns['id'] == link_id)
    .values(url=form.url.data, switch_date=form.switch_date.data)

if form.password.data:
    salt = os.urandom(32)
    password = hashlib.pbkdf2_hmac(
        'sha256',
        form.password.data.encode('utf-8'),
        salt,
        100000)

    link_update = link_update.values(password=password)

    link_select = await conn.execute(table.select().where(table.columns['id'] == link_id))
    link_data = await link_select.fetchone()
    if link_data.password:
        salt_query = salts.update()
            .where(salts.columns['identifier'] == link_data.identifier)
            .values(salt=fresh_salt))
    else:
        salt_query = salts.insert()
            .values(identifier=link_data.identifier, salt=salt))

await conn.execute(link_update)
await conn.execute(salt_query)
```

but it's of course very subjective... What do you think? Do you prefer your solution?

Reply:
That looks nicer, thanks. I had to change it a little bit to make it work, but hopefully it still looks acceptable.